### DISCUSSION TITLE
How to get session token of login API when using pact ?

### DISCUSSION THREAD

**Discussion by LynnJing** [pact-foundation/pact-python#628] [Q&A]
Hi, 
I want to use pact to do contract testing. My case is as below. 
Step1 : Use Login to get related token 
Step2: Call get API using above token to generate Pact json file 

Could you please have a look about this and give me one example for above Case?

Thank you very much

**Comment by JP-Ellis** [comment]
Hi Lynn,

I'm not entirely how Pact is meant to fit in with the above example. As Pact is intended to mock the other side of the interaction (be it the consumer or provider), you shouldn't need to login though you may require that a authorization header be set.

For example, you might have a Pact which, in English, would look like:

> Upon receiving `GET /some/api`
> With header `Authorization: Bearer 123abc`
> Will respond with `200`
> With body `{"some": "result"}`

**Reply by LynnJing** [reply]
Thanks for your support on this. I tried to add related auth token in header in Python V3, However, which throw 

> Could not find key "csp-auth-token" (keys present are: Content-Length, Content-Type, Host, User-Agent, Accept-Encoding, Accept, Connection, Version) at $.headers" this error.

Is this because Python V3 cannot support self-defined header? 

Could you please support to have a look about this?

Related Code as below

```python
     def test_csp_org(self):
        help=Helper()
        token=help.get_csp_auth()
        headers = {"csp-auth-token": ""}
        headers["csp-auth-token"]=token
        (
            pact
                .given('test get csp org')
                .upon_receiving(contract_name)
                .with_request(method='get', headers=headers, path=csp_org_url)
                .will_respond_with(status=200, headers=headers)
        )
        # mock provider,pact's mock server,default port is 1234
        with pact:
            resp = requests.get('http://localhost:1234{}'.format(csp_org_url), json=p).json()
            print(resp)
            pact.verify()
```

Thank you very much, 
Lynn

**Reply by mefellows** [reply]
This is because you have said you expect your API client to send a request with that header, but none was received.

I believe you have worked that out by marking off the issue, but wanted to confirm here for future users passing this discussion by.

**Reply by LynnJing** [reply]
Yes, thank you very much, mefellows. 
After I set the same header at

```python
with self.pact:
    resp = requests.get(
        'http://localhost:1234{}'.format(csp_org_url),
        headers=headers,
    )
    ...
```

which can work now.

**Reply by mefellows** [reply]
No worries!

Just for posterity, usually in your Pact test you would use your API client instead of a general API library (such as `requests`), as its a unit test of that piece of code that is the purpose of a Pact test.

You may know this and it might just be a test, but I wanted to make that point publicly visible.

**Comment by mefellows** [comment]
This is a common misunderstanding about how Pact and more generally, contract testing, is supposed to work.

In Pact, each test should be isolated and independent from other requests. So you should not have to sequence calls this way - this is really a functional test. From our docs:

> Each interaction in a pact should be verified in isolation, with no context maintained from the previous interactions. Tests that depend on the outcome of previous tests are brittle and land you back in integration test hell, which is the nasty place you’re trying to escape by using pacts.

As @JP-Ellis mentioned, you can use [provider states](https://docs.pact.io/getting_started/provider_states) to specify the pre-condition for the test, and use https://docs.pact.io/provider/handling_auth to work out the best strategy to manage the provider verification side of things.
